### PR TITLE
BUG: Fix rolling median and quantile with closed='left' and closed='neither' (#26005)

### LINKED ISSUE #26005: offset-based rolling window, multiple issues with closed='left'
#### Code Sample

```python
# Case 1: single row
df1 = pd.DataFrame({'B': [0]}, index=[pd.Timestamp('20130101 09:00:00')])
df1.rolling('1s', closed='left').median()  # <- raises 'MemoryError: skiplist_init failed'

# Case 2: multiple rows, but entries separated by a larger time than the specified window
df2 = pd.DataFrame({'B': [0, 1]}, index=[pd.Timestamp('20130101 09:00:00'), pd.Timestamp('20130101 09:00:02')])
df2.rolling('1s', closed='left').median() # <- raises 'MemoryError: skiplist_init failed'
df2.rolling('1s', closed='left').max() # <- no error, but second entry seems incorrect

# Case 3: as long as at least one row has other entries in its window, it runs without 
# an exception but the values are suspect
df3 = pd.DataFrame({'B': [1, 2, 3]}, index=[pd.Timestamp('20130101 09:00:00'), pd.Timestamp('20130101 09:00:02'), pd.Timestamp('20130101 09:00:03')])
df3.rolling('1s', closed='left').median() # <- no exception, but the values seem incorrect
df3.rolling('1s', closed='left').max() # ditto
df3.rolling('2s', closed='left').median() # ditto (note longer window)
```
#### Problem description

Obviously, the exception cases are a big problem and should be addressed. The other cases laid out here seem to give unexpected results that are inconsistent with other aggregations (such as mean and sum) that do seem to be operating correctly. Note that using closed='right' or closed='both' does seem to give results consistent with my expectations, while using closed='neither' yields similar problems as closed='left'. (So, it would seem that the common factor here is whether or not the input rows are included in their own rolling windows.)

#### Expected Output

Case 1: since there are no other entries in the input row's window, I would expect that the median aggregation return NaN. (This would be consistent with mean, max, etc. for this case.)

```python
                      B
2013-01-01 09:00:00 NaN
```

Case 2: since neither input row should have any other entries in their windows, I would expect that the median and max results should all be NaN. (This would be consistent with what the mean aggregation returns for this case.)
```python
                     B
2013-01-01 09:00:00  NaN
2013-01-01 09:00:02  NaN
```

Case 3a and 3b (1s window): since neither of the first two input rows should have any other entries in their windows, I would expect that their median and max results should all be NaN. since the last row does have an entry in its window (the second row) I would expect that both the median and max should be 2.0. (This would be consistent with what the mean aggregation returns for this case.)
```python
                     B
2013-01-01 09:00:00  NaN
2013-01-01 09:00:02  NaN
2013-01-01 09:00:03  2.0
```
Case 3c (2s window): since the first row should have no entries in its window, I would expect the first output row to be NaN. the second row will have the first entry in its window, so I would expect its output to be 1.0. similarly, the last row will have the second entry in its window and I would expect its output to be 2.0.
```python
                     B
2013-01-01 09:00:00  NaN
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  2.0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-17134-Microsoft
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None
pandas: 0.24.2
pytest: None
pip: 19.0.1
setuptools: 40.6.3
Cython: None
numpy: 1.16.0
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.3.0
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by ihsansecer**
- [x] Fixes issues with rolling `median` (and also `quantile`) `closed='left'` and `closed='neither'` in #26005 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Here is my understanding of how it should work:

*Same as test cases, there is one day between each element and window size is 3 days.

For `closed='both'` (**()** denotes Nth window and **-** is just to state window is out of ):
```
N=0 | (-, -, - [0), 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=1 | -, (-, - [0, 1), 2, 3, 4, 5, 6, 7, 8, 9]
N=2 | -, -, (- [0, 1, 2), 3, 4, 5, 6, 7, 8, 9]
N=3 | -, -, - [(0, 1, 2, 3), 4, 5, 6, 7, 8, 9]
...
```

For `closed='right'`:
```
N=0 | -, (-, - [0), 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=1 | -, -, (- [0, 1), 2, 3, 4, 5, 6, 7, 8, 9]
N=2 | -, -, - [(0, 1, 2), 3, 4, 5, 6, 7, 8, 9]
N=3 | -, -, - [0, (1, 2, 3), 4, 5, 6, 7, 8, 9]
...
```
For `closed='left'`:
```
N=0 | (-, -, -) [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=1 | -, (-, - [0), 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=2 | -, -, (- [0, 1), 2, 3, 4, 5, 6, 7, 8, 9]
N=3 | -, -, - [(0, 1, 2), 3, 4, 5, 6, 7, 8, 9]
...
```

For `closed='neither'`:
```
N=0 | -, (-, -) [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=1 | -, -, (- [0), 1, 2, 3, 4, 5, 6, 7, 8, 9]
N=2 | -, -, - [(0, 1), 2, 3, 4, 5, 6, 7, 8, 9]
N=3 | -, -, - [0, (1, 2), 3, 4, 5, 6, 7, 8, 9]
...
```

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thank @ihsansecer. It looks like #26005 reported multiple issues. Do your changes fix the memory skiplist issues? If so, can you add tests / release note? Otherwise, edit the original post to remove the "closes", so that it remains open.

cc @WillAyd

File: pandas/tests/test_window.py [hunk 1]
```diff
@@ -594,6 +594,20 @@ def test_closed_min_max_minp(self, func, closed, expected):
         expected = pd.Series(expected, index=ser.index)
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("closed,expected", [
+        ('right', [0, 0.5, 1, 2, 3, 4, 5, 6, 7, 8]),
```
Comment (TomAugspurger) [inline pandas/tests/test_window.py]: Can you walk me through why this is expected for `closed='right'`? For some reason, this looks like `closed='left'` to me, but I'm probably missing something.

**Comment by ihsansecer** [comment]
@TomAugspurger not yet checked issues with memory skiplist and `max` function but I am planning to have a look and open another PR. Thanks

File: pandas/tests/test_window.py (hunk 1 above)
Comment (ihsansecer) [reply pandas/tests/test_window.py]: Just wanted to enclose other options for closed argument. `left` and `neither` options are causing issues as you stated. Could be removed if unnecessary

Comment (jreback) [reply pandas/tests/test_window.py]: can you should the results of all 4 of these cases *in* the PR itself (you can update the top is good), just to get an easy visual on the results.

**Review by jreback — CHANGES_REQUESTED** [review]
comments

File: pandas/tests/test_window.py (hunk 1 above)
Comment (ihsansecer) [reply pandas/tests/test_window.py]: @TomAugspurger Sorry, I just realize I misread your comment. @jreback @TomAugspurger  I tried to explain all 4 cases at the top of PR.

**Comment by ihsansecer** [comment]
It seems `quantile` is also producing unexpected results same as stated in the issue. If there is no problem with this fix I will apply the same to `quantile` function.

**Comment by jreback** [comment]
yep, can you merge master and push again here.

**Comment by ihsansecer** [comment]
@jreback Merged and changed `roll_quantile` as mentioned

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ihsansecer nice patch!

**Pull request merged**